File: app/src/main/java/jp/gr/java_conf/foobar/testmaker/service/views/ImageTextButton.java
package jp.gr.java_conf.foobar.testmaker.service.views;

import android.content.Context;
import android.content.res.TypedArray;
import android.util.AttributeSet;
import android.util.Log;
import android.view.KeyEvent;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.widget.ImageButton;
import android.widget.LinearLayout;
import android.widget.TextView;

import jp.gr.java_conf.foobar.testmaker.service.R;

/**
 * Created by keita on 2016/07/11.
 */
public class ImageTextButton extends LinearLayout {

    private OnClickListener listener;

    ImageButton button;

    public ImageTextButton(Context context) {
        super(context);
    }

    public ImageTextButton(Context context, AttributeSet attrs) {
        super(context, attrs);

        TypedArray a = context.obtainStyledAttributes(attrs,
                R.styleable.ImageTextButton);

        View layout = LayoutInflater.from(context).inflate(R.layout.image_text_button, this);

        button = layout.findViewById(R.id.button);
        button.setImageDrawable(a.getDrawable(R.styleable.ImageTextButton_image));

        TextView text = layout.findViewById(R.id.text);

        if(a.getText(R.styleable.ImageTextButton_text)!=null){
            if(!a.getText(R.styleable.ImageTextButton_text).equals("")) {
                text.setVisibility(VISIBLE);
                text.setText(a.getText(R.styleable.ImageTextButton_text));


            }
        }

        a.recycle();
    }

    public ImageTextButton(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    @Override
    public boolean dispatchTouchEvent(MotionEvent event) {
        if (event.getAction() == MotionEvent.ACTION_UP) {
            if (listener != null){
                button.setPressed(true);
                listener.onClick(this);
            }
        }
        return super.dispatchTouchEvent(event);
    }

    @Override
    public boolean dispatchKeyEvent(KeyEvent event) {
        if (event.getAction() == KeyEvent.ACTION_UP
                && (event.getKeyCode() == KeyEvent.KEYCODE_DPAD_CENTER
                || event.getKeyCode() == KeyEvent.KEYCODE_ENTER)) {
            if (listener != null){
                button.setPressed(true);
                listener.onClick(this);
            }
        }
        return super.dispatchKeyEvent(event);
    }

    public void setOnClickListener(OnClickListener listener) {
        this.listener = listener;
    }


}

File: app/src/main/java/jp/gr/java_conf/foobar/testmaker/service/view/share/ImageTextButton.kt
package jp.gr.java_conf.foobar.testmaker.service.view.share

import android.content.Context
import android.util.AttributeSet
import android.view.KeyEvent
import android.view.LayoutInflater
import android.view.MotionEvent
import android.view.View
import android.widget.LinearLayout
import jp.gr.java_conf.foobar.testmaker.service.R
import kotlinx.android.synthetic.main.image_text_button.view.*

/**
 * Created by keita on 2016/07/11.
 */
class ImageTextButton : LinearLayout {

    private var listener: View.OnClickListener? = null

    constructor(context: Context) : super(context)

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {

        val a = context.obtainStyledAttributes(attrs,
                R.styleable.ImageTextButton)

        button.setImageDrawable(a.getDrawable(R.styleable.ImageTextButton_image))

        if (a.getText(R.styleable.ImageTextButton_text) != null) {

            if (a.getText(R.styleable.ImageTextButton_text) != "") {

                text.visibility = View.VISIBLE
                text.text = a.getText(R.styleable.ImageTextButton_text)

            }
        }

        a.recycle()
    }

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr)

    init { LayoutInflater.from(context).inflate(R.layout.image_text_button, this) }

    override fun dispatchTouchEvent(event: MotionEvent): Boolean {
        if (event.action == MotionEvent.ACTION_UP) {
            if (listener != null) {
                button.isPressed = true
                listener!!.onClick(this)
            }
        }
        return super.dispatchTouchEvent(event)
    }

    override fun dispatchKeyEvent(event: KeyEvent): Boolean {
        if (event.action == KeyEvent.ACTION_UP && (event.keyCode == KeyEvent.KEYCODE_DPAD_CENTER || event.keyCode == KeyEvent.KEYCODE_ENTER)) {
            if (listener != null) {
                button.isPressed = true
                listener!!.onClick(this)
            }
        }
        return super.dispatchKeyEvent(event)
    }

    override fun setOnClickListener(listener: View.OnClickListener?) {
        this.listener = listener
    }


}
